proof/compress-p1.saw

```diff
@@ -0,0 +1,201 @@
+
+    // SAW struggles to reason about `HTCE1::sgn0 (affinified.1)`, but we don't
+    // need it anyway so this spec only specifies the sign_F_p bit.
+    ret <- crucible_fresh_var "ret" limb_type;
+    crucible_postcond {{ (ret!1) == (sign_F_p affinified.1) }};
+    crucible_return (crucible_term ret);
+  };
+};
+
+let POINTonE1_Compress_BE_proof = do {
+     // Unfolding needed for fp rewrites to apply
+     unfolding ["POINTonE1_affine_abs", "POINTonE1_affine_rep", "POINTonE1_invariant"];
+     (apply_fp_rewrites);
+     simplify (addsimp one_mont_p_abs_thm empty_ss);
+     w4_unint_z3 ["POINTonE1_abs", "POINTonE1_invariant", "Fp", "fp_abs", "I2OSP48"];
+  };
+
+// Prove POINTonE1_Compress_BE matches the spec for both projective and affine
+// points.
+POINTonE1_Compress_BE_projective_ov <- custom_verify "POINTonE1_Compress_BE"
+  [POINTonE1_from_Jacobian_ov, POINTonE1_affine_Compress_BE_ov, vec_is_equal_fp_ov]
+  (POINTonE1_Compress_BE_spec false)
+  (POINTonE1_Compress_BE_proof);
+POINTonE1_Compress_BE_affine_ov <- custom_verify "POINTonE1_Compress_BE"
+  [POINTonE1_affine_Compress_BE_ov, vec_is_equal_fp_ov]
+  (POINTonE1_Compress_BE_spec true)
+  (POINTonE1_Compress_BE_proof);
+
+
+let blst_p1_compress_spec is_point_affine = do {
+  out_ptr <- crucible_alloc (llvm_array 48 (llvm_int 8));
+  (inp, inp_ptr) <- ptr_to_fresh_readonly "in" POINTonE1_type;
+  crucible_precond {{ POINTonE1_invariant inp }};
+
+  if is_point_affine then do {
+    // Point is affine, so Z must be 1
+    crucible_precond {{ (fp_abs (inp.2)) == Fp.field_unit }};
+    crucible_precond {{ (fp_abs (inp.2)) != Fp.field_zero }};
+
+    // Have the "Z=0 <=> infinity" issue again; see the notes in `map_to_g1.saw`.
+    // So, passing in a value (0,0,1) would give a result not agreeing with
+    // the specification.
+    crucible_precond {{ ~(is_point_O E (fp_abs inp.0, fp_abs inp.1)) }};
+  } else do {
+    // Point is projective, so Z must not be 1
+    crucible_precond {{ (fp_abs (inp.2)) != Fp.field_unit }};
+    crucible_precond {{ is_point_projective E (POINTonE1_abs inp) }}; // on the curve
+  };
+
+  crucible_execute_func [out_ptr, inp_ptr];
+  if is_point_affine then do {
+    crucible_points_to out_ptr (crucible_term {{ serialize_E1 (fp_abs inp.0, fp_abs inp.1) }});
+  } else do {
+    crucible_points_to out_ptr (crucible_term {{ serialize_E1 (affinify E (POINTonE1_abs inp)) }});
+  };
+};
+
+// This is the conditional rule showing that the Z coordinate check does the right thing
+// This may be provable in SAW but for now we assume it as an algebraic fact.
+is_point_O_thm'' <- admit_cryptol (rewrite (cryptol_ss())
+  {{ \p -> peq (fp_abs(p.2)) (Fp.field_zero) ==
+           if is_point_projective E (POINTonE1_abs p)
+           then is_point_O E (affinify E (POINTonE1_abs p))
+           else apply peq (fp_abs(p.2)) (Fp.field_zero) }});
+
+// A reversed version of is_point_O_thm''
+is_point_O_rev_thm <- admit_cryptol (rewrite (cryptol_ss())
```

Comment:
This could in principle be obtained by rewriting with the previous rule (modulo the fact that one uses `peq ...` and the other `is_zero ...`)